src/Microsoft.PowerShell.Commands.Diagnostics/PdhHelper.cs

```diff
@@ -542,15 +542,16 @@ public uint ConnectToDataSource(StringCollection blgFileNames)
                 return ConnectToDataSource(blgFileNames[0]);
             }
 
-            string doubleNullTerminated = string.Empty;
-            foreach (string fileName in blgFileNames)
+            StringBuilder doubleNullTerminated = new StringBuilder();
+            for (int i = 0; i < blgFileNames.Count; i++)
             {
-                doubleNullTerminated += fileName + '\0';
+                string fileName = blgFileNames[i];
+                doubleNullTerminated.Append(fileName).Append('\0');
             }
 
-            doubleNullTerminated += '\0';
+            doubleNullTerminated.Append('\0');
 
-            return ConnectToDataSource(doubleNullTerminated);
+            return ConnectToDataSource(doubleNullTerminated.ToString());
```

Comment:
I would assume the string allocation is a much bigger issue than the loop enumerator. In which case:

```csharp
var dataString = new StringBuilder();
foreach (string fileName in blgFileNames)
{
    dataString.Append(fileName).Append('\0');
}
dataString.Append('\0');

return ConnectToDataSource(dataString);
```

Reply:
But how can we avoid the string allocation?

Reply:
As in preventing allocating a string altogether? It seems to depend heavily on the requirements of the input and output.

If the input is a fixed, known size in advance, `string.Create()` might be useful.

The output requires a rather idiosyncratic string according to the [docs](https://docs.microsoft.com/en-us/windows/win32/api/pdh/nf-pdh-pdhbindinputdatasourcew#parameters), so we are forced to allocate one.